# Laptop Market Analysis

## Dataset Overview

This project analyzes a dataset of laptops collected from an online electronics store.

The main goal is to explore laptop prices and understand how they are related to hardware specifications such as RAM, storage, processor, graphics, and display characteristics.

In [5]:
import pandas as pd

In [6]:
df = pd.read_csv('../data/laptops.csv')

In [7]:
df.head()

,Title,Price,Image URL,Specifications
0,"Laptop HP OmniBook 3 16-by0028nwx 16"" R5 40 8G...",1 999.00 zł,https://f00.esfr.pl/foto/4/192315626929/e8ea62...,"{'Ekran': '16 "", 1920 x 1200 pikseli', 'Proce..."
1,"Laptop gamingowy Lenovo LOQ 15IAX9 15,6"" 144Hz...",3 499.00 zł,https://f00.esfr.pl/foto/7/177142838209/75780b...,"{'Ekran': '15,6 "", 1920 x 1080 pikseli 144 Hz..."
2,Laptop gamingowy Dell Alienware 16 Aurora AC16...,3 299.00 zł,https://f00.esfr.pl/foto/2/188164628089/f7bd6d...,"{'Ekran': '16 "", 2560 x 1600 pikseli 120 Hz',..."
3,Laptop gamingowy ASUS TUF Gaming F16 FX607VJB-...,3 299.00 zł,https://f00.esfr.pl/foto/4/191218243201/6365f8...,"{'Ekran': '16 "", 1920 x 1200 pikseli 144 Hz',..."
4,"Laptop Lenovo IdeaPad Slim 3 16ARP10 16"" R5 75...",2 699.00 zł,https://f00.esfr.pl/foto/4/187619326041/eed87c...,"{'Ekran': '16 "", 1920 x 1200 pikseli 60 Hz', ..."


### Dataset Structure

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1075 entries, 0 to 1074
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Title           1075 non-null   object
 1   Price           842 non-null    object
 2   Image URL       1075 non-null   object
 3   Specifications  1075 non-null   object
dtypes: object(4)
memory usage: 33.7+ KB


### Missing Values

In [9]:
missing_values = pd.DataFrame({
    'Missing': df.isna().sum(),
    'Percent': (df.isna().mean() * 100).round(2)
})

missing_values

,Missing,Percent
Title,0,0.00
Price,233,21.67
Image URL,0,0.00
Specifications,0,0.00


Price is the only column with missing values. There are 233 missing price records out of 1,075 laptops (21.67%).

### Duplicate Rows

In [10]:
df.duplicated().sum()

0

No fully duplicated rows were found in the dataset.

### Initial Observations

The dataset contains 1,075 laptop records and 4 columns.

There are 233 missing prices and no fully duplicated rows:

- `Title` — product name
- `Price` — product price
- `Image URL` — product image URL
- `Specifications` — hardware and display specifications stored as text

`Price` is currently stored as text and will require cleaning before numerical analysis.

## Data Cleaning

Before performing the analysis, the raw data needs to be cleaned and transformed into a format suitable for analysis.

The main cleaning tasks are:

- convert `Price` from text to numeric values
- parse the `Specifications` column
- Identify and standardize inconsistent values

### Clean Price

The `Price` column is stored as text and includes the `zł` currency symbol and formatting characters.

Before converting it to a numeric type, I will inspect several raw values.

In [11]:
df["Price"].head(10)

0    1 999.00 zł
1    3 499.00 zł
2    3 299.00 zł
3    3 299.00 zł
4    2 699.00 zł
5    4 222.00 zł
6    3 199.00 zł
7    4 999.00 zł
8    3 099.00 zł
9    2 499.00 zł
Name: Price, dtype: object

In [12]:
df['Price'] = df['Price'].str.replace('zł', '').str.replace(' ', '').astype(float)

In [13]:
df["Price"].head(10)

0    1999.0
1    3499.0
2    3299.0
3    3299.0
4    2699.0
5    4222.0
6    3199.0
7    4999.0
8    3099.0
9    2499.0
Name: Price, dtype: float64

In [14]:
df["Price"].isna().sum()

233

In [15]:
df["Price"].describe()

count      842.000000
mean      6219.193147
std       4058.497990
min       1299.000000
25%       3661.250000
50%       4999.000000
75%       7199.000000
max      27999.000000
Name: Price, dtype: float64

The dataset contains 842 laptops with available prices. The median price is lower than the mean price, which suggests that the price distribution is right-skewed.

### Laptops Without Price

Let's look at laptops where the price is missing.

In [16]:
df_no_price = df[df['Price'].isna()]
df_no_price.head(10)

,Title,Price,Image URL,Specifications
842,Laptop gamingowy Lenovo Legion Pro 5 16IAX10 O...,NaN,https://f00.esfr.pl/foto/6/187619370057/144499...,"{'Ekran': '16 "", 2560 x 1600 pikseli 165 Hz',..."
843,Laptop gamingowy Gigabyte Gaming A16 3THK3EE89...,NaN,https://f00.esfr.pl/foto/6/173811870289/f5110a...,"{'Ekran': '16 "", 1920 x 1200 pikseli 165 Hz',..."
844,Laptop 2w1 HP OmniBook Ultra Flip 14-fh0216nw ...,NaN,https://f00.esfr.pl/foto/3/184148947905/cfd7ff...,"{'Ekran': '14 "", ekran dotykowy 2880 x 1800 pi..."
845,Laptop gamingowy MSI Titan 18 HX Dragon Editio...,NaN,https://f00.esfr.pl/foto/7/192277966257/974670...,"{'Ekran': '18 "", 3840 x 2400 pikseli 240 Hz',..."
846,"Laptop gamingowy HP Victus 15-fb3078nw 15,6"" 1...",NaN,https://f00.esfr.pl/foto/5/184148874433/93b35a...,"{'Ekran': '15,6 "", 1920 x 1080 pikseli 144 Hz..."
847,"Apple MacBook Air 2025 15,3"" M4 16GB RAM 512GB...",NaN,https://f00.esfr.pl/foto/5/165147977201/1cydbu...,"{'Ekran': '15,3"", 2880 x 1864 pikseli', 'Proc..."
848,Laptop gamingowy Gigabyte Gaming A18 3VHK3EEF9...,NaN,https://f00.esfr.pl/foto/1/173812021065/4ba17a...,"{'Ekran': '18 "", 1920 x 1200 pikseli 165 Hz',..."
849,"Apple MacBook Air 2025 15,3"" M4 16GB RAM 512GB...",NaN,https://f00.esfr.pl/foto/9/165147820657/4tivv8...,"{'Ekran': '15,3"", 2880 x 1864 pikseli', 'Proc..."
850,"Apple MacBook Air 2025 15,3"" M4 24GB RAM 512GB...",NaN,https://f00.esfr.pl/foto/7/165147723473/1dv93f...,"{'Ekran': '15,3"", 2880 x 1864 pikseli', 'Proc..."
851,"Apple MacBook Air 2025 13,6"" M4 24GB RAM 512GB...",NaN,https://f00.esfr.pl/foto/2/159937723345/10qx7y...,"{'Ekran': '13,6 "", 2560 x 1664 pikseli', 'Pro..."


### Final Check

Let's check the final dataset after the price conversion.

In [17]:
df.shape

(1075, 4)

In [18]:
df["Price"].dtype

dtype('float64')

In [19]:
df["Price"].isna().sum()

233

In [20]:
df.duplicated().sum()

0

### Laptops With Known Prices

I will create a separate DataFrame for laptops with known prices.
I will keep the original DataFrame unchanged.

In [21]:
df_price = df.dropna(subset=["Price"])

In [22]:
df_price.shape

(842, 4)

In [23]:
df_price["Price"].isna().sum()

0

In [24]:
df.shape

(1075, 4)

### Final Check

After cleaning the price column:

The dataset still contains 1,075 rows.

The `Price` column is `float64`.

There are 233 missing prices.

There are no fully duplicated rows.

The `df_price` DataFrame contains only laptops with a known price.

Some laptop titles appear more than once, so I will check these records later.

### Parse Specifications

The `Specifications` column contains product specifications stored as text representations of Python dictionaries.

To analyze these specifications separately, I will first convert the text into dictionaries and then expand the dictionary keys into individual columns.

In [25]:
df["Specifications"].dtype

dtype('O')

In [26]:
type(df["Specifications"].iloc[0])

str

In [27]:
df["Specifications"].iloc[0]

'{\'Ekran\': \'16 ",  1920 x 1200 pikseli\', \'Procesor\': \'AMD Ryzen™ 5 40\', \'Pamięć\': \'8 GB  LPDDR5 5500 Mhz RAM\', \'Grafika\': \'AMD  Radeon™ 610M Graphics\', \'Dysk\': \'256 GB SSD\', \'System operacyjny\': \'Windows 11 Home Edition\'}'

The values are stored as strings, so I will convert them into Python dictionaries.

In [28]:
import ast

In [29]:
df["Specifications"] = df["Specifications"].apply(ast.literal_eval)

In [30]:
df["Specifications"].apply(type).value_counts()

Specifications
<class 'dict'>    1075
Name: count, dtype: int64

In [31]:
df["Specifications"].head()

0    {'Ekran': '16 ",  1920 x 1200 pikseli', 'Proce...
1    {'Ekran': '15,6 ",  1920 x 1080 pikseli 144 Hz...
2    {'Ekran': '16 ",  2560 x 1600 pikseli 120 Hz',...
3    {'Ekran': '16 ",  1920 x 1200 pikseli 144 Hz',...
4    {'Ekran': '16 ",  1920 x 1200 pikseli 60 Hz', ...
Name: Specifications, dtype: object

The `Specifications` values are now Python dictionaries.

I will use `json_normalize()` to expand the dictionaries into separate columns.

In [32]:
specs = pd.json_normalize(df["Specifications"])

In [33]:
specs.head()

,Ekran,Procesor,Pamięć,Grafika,Dysk,System operacyjny
0,"16 "", 1920 x 1200 pikseli",AMD Ryzen™ 5 40,8 GB LPDDR5 5500 Mhz RAM,AMD Radeon™ 610M Graphics,256 GB SSD,Windows 11 Home Edition
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",Intel® Core™ i5 12gen 12600HX,16 GB DDR5 4800 Mhz RAM,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,512 GB SSD,Windows 11 Home Edition
2,"16 "", 2560 x 1600 pikseli 120 Hz",Intel® Core™ 5 15gen 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
3,"16 "", 1920 x 1200 pikseli 144 Hz",Intel® Core™ 5 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
4,"16 "", 1920 x 1200 pikseli 60 Hz",AMD Ryzen™ 5 7535HS,24 GB DDR5 4800 Mhz RAM,AMD Radeon™ 660M Graphics,512 GB SSD,Windows 11 Home Edition


In [34]:
specs.shape

(1075, 6)

In [35]:
specs.columns

Index(['Ekran', 'Procesor', 'Pamięć', 'Grafika', 'Dysk', 'System operacyjny'], dtype='object')

I will rename the columns to make them easier to work with.

In [36]:
specs = specs.rename(columns={
    "Ekran": "screen",
    "Procesor": "processor",
    "Pamięć": "ram",
    "Grafika": "graphics",
    "Dysk": "storage",
    "System operacyjny": "os"
})

In [37]:
specs.head()

,screen,processor,ram,graphics,storage,os
0,"16 "", 1920 x 1200 pikseli",AMD Ryzen™ 5 40,8 GB LPDDR5 5500 Mhz RAM,AMD Radeon™ 610M Graphics,256 GB SSD,Windows 11 Home Edition
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",Intel® Core™ i5 12gen 12600HX,16 GB DDR5 4800 Mhz RAM,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,512 GB SSD,Windows 11 Home Edition
2,"16 "", 2560 x 1600 pikseli 120 Hz",Intel® Core™ 5 15gen 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
3,"16 "", 1920 x 1200 pikseli 144 Hz",Intel® Core™ 5 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
4,"16 "", 1920 x 1200 pikseli 60 Hz",AMD Ryzen™ 5 7535HS,24 GB DDR5 4800 Mhz RAM,AMD Radeon™ 660M Graphics,512 GB SSD,Windows 11 Home Edition


In [38]:
specs.isna().sum()

screen       0
processor    0
ram          0
graphics     0
storage      0
os           0
dtype: int64

In [39]:
df = pd.concat([df.drop(columns="Specifications"), specs], axis=1)

### Check the New Structure

The dataset now contains separate columns for the main laptop specifications.

In [40]:
df.head()

,Title,Price,Image URL,screen,processor,ram,graphics,storage,os
0,"Laptop HP OmniBook 3 16-by0028nwx 16"" R5 40 8G...",1999.0,https://f00.esfr.pl/foto/4/192315626929/e8ea62...,"16 "", 1920 x 1200 pikseli",AMD Ryzen™ 5 40,8 GB LPDDR5 5500 Mhz RAM,AMD Radeon™ 610M Graphics,256 GB SSD,Windows 11 Home Edition
1,"Laptop gamingowy Lenovo LOQ 15IAX9 15,6"" 144Hz...",3499.0,https://f00.esfr.pl/foto/7/177142838209/75780b...,"15,6 "", 1920 x 1080 pikseli 144 Hz",Intel® Core™ i5 12gen 12600HX,16 GB DDR5 4800 Mhz RAM,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,512 GB SSD,Windows 11 Home Edition
2,Laptop gamingowy Dell Alienware 16 Aurora AC16...,3299.0,https://f00.esfr.pl/foto/2/188164628089/f7bd6d...,"16 "", 2560 x 1600 pikseli 120 Hz",Intel® Core™ 5 15gen 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
3,Laptop gamingowy ASUS TUF Gaming F16 FX607VJB-...,3299.0,https://f00.esfr.pl/foto/4/191218243201/6365f8...,"16 "", 1920 x 1200 pikseli 144 Hz",Intel® Core™ 5 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
4,"Laptop Lenovo IdeaPad Slim 3 16ARP10 16"" R5 75...",2699.0,https://f00.esfr.pl/foto/4/187619326041/eed87c...,"16 "", 1920 x 1200 pikseli 60 Hz",AMD Ryzen™ 5 7535HS,24 GB DDR5 4800 Mhz RAM,AMD Radeon™ 660M Graphics,512 GB SSD,Windows 11 Home Edition


In [41]:
df.isna().sum()

Title          0
Price        233
Image URL      0
screen         0
processor      0
ram            0
graphics       0
storage        0
os             0
dtype: int64

In [42]:
df.shape

(1075, 9)

### Result

The dataset now contains 1,075 rows and 9 columns.

The laptop specifications are stored in separate columns.

There are 233 missing prices, but the specification columns do not contain missing values.

The specification values are still stored as text. I will clean and extract useful numeric values in the next step.

## Clean Text Columns

In [43]:
cols = ['Title', 'screen', 'processor', 'ram', 'graphics', 'storage', 'os']

df[cols] = df[cols].apply(lambda col: col.str.strip())

In [44]:
(df[cols] == "").sum()

Title        0
screen       0
processor    0
ram          0
graphics     0
storage      0
os           0
dtype: int64

No empty strings (`""`) were found in the specification columns.

In [45]:
df[['os', 'ram', 'storage']].value_counts()

os                       ram                          storage   
Windows 11 Professional  16 GB  DDR5 5600 Mhz RAM     512 GB SSD    54
Windows 11 Home Edition  16 GB  DDR5 5600 Mhz RAM     512 GB SSD    48
                         16 GB  DDR4 3200 MHz RAM     512 GB SSD    45
                         16 GB  DDR5 RAM              512 GB SSD    42
                         16 GB  LPDDR5X RAM           512 GB SSD    27
                                                                    ..
Windows 11 Professional  16 GB  LPDDR5x 8000 Mhz RAM  1 TB SSD       1
Windows 11 Home Edition  24 GB  LPDDR5x 8533 Mhz RAM  512 GB SSD     1
Windows 11 Professional  16 GB  LPDDR5x 7467 Mhz RAM  512 GB SSD     1
                         16 GB  LPDDR5X RAM           512 GB SSD     1
Chrome OS                16 GB  LPDDR5X RAM           512 GB SSD     1
Name: count, Length: 249, dtype: int64

Check how often different combinations of operating system, RAM, and storage occur.

The results show 249 unique combinations. Some combinations are common, while others occur only once.

In [46]:
df.duplicated().sum()

0

In [47]:
df["Title"].duplicated().sum()

18

In [48]:
df.head()

,Title,Price,Image URL,screen,processor,ram,graphics,storage,os
0,"Laptop HP OmniBook 3 16-by0028nwx 16"" R5 40 8G...",1999.0,https://f00.esfr.pl/foto/4/192315626929/e8ea62...,"16 "", 1920 x 1200 pikseli",AMD Ryzen™ 5 40,8 GB LPDDR5 5500 Mhz RAM,AMD Radeon™ 610M Graphics,256 GB SSD,Windows 11 Home Edition
1,"Laptop gamingowy Lenovo LOQ 15IAX9 15,6"" 144Hz...",3499.0,https://f00.esfr.pl/foto/7/177142838209/75780b...,"15,6 "", 1920 x 1080 pikseli 144 Hz",Intel® Core™ i5 12gen 12600HX,16 GB DDR5 4800 Mhz RAM,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,512 GB SSD,Windows 11 Home Edition
2,Laptop gamingowy Dell Alienware 16 Aurora AC16...,3299.0,https://f00.esfr.pl/foto/2/188164628089/f7bd6d...,"16 "", 2560 x 1600 pikseli 120 Hz",Intel® Core™ 5 15gen 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
3,Laptop gamingowy ASUS TUF Gaming F16 FX607VJB-...,3299.0,https://f00.esfr.pl/foto/4/191218243201/6365f8...,"16 "", 1920 x 1200 pikseli 144 Hz",Intel® Core™ 5 210H,16 GB DDR5 5600 Mhz RAM,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,512 GB SSD,Windows 11 Home Edition
4,"Laptop Lenovo IdeaPad Slim 3 16ARP10 16"" R5 75...",2699.0,https://f00.esfr.pl/foto/4/187619326041/eed87c...,"16 "", 1920 x 1200 pikseli 60 Hz",AMD Ryzen™ 5 7535HS,24 GB DDR5 4800 Mhz RAM,AMD Radeon™ 660M Graphics,512 GB SSD,Windows 11 Home Edition


In [49]:
df.isna().sum()

Title          0
Price        233
Image URL      0
screen         0
processor      0
ram            0
graphics       0
storage        0
os             0
dtype: int64

In [50]:
df.shape

(1075, 9)

### Cleaning Notes

The text columns were cleaned by removing extra spaces.

There are no empty strings in the main text columns.

There are no fully duplicated rows.

There are 18 repeated laptop titles, but these rows will not be removed because the same laptop can have different prices or product data.

There are still 233 missing prices. These values will be kept for now.

### Extract Numeric Features

The specification columns are now separated, but their values are still stored as text.

To make the data suitable for numerical analysis, I will extract relevant numeric features from the `screen`, `ram`, and `storage` columns.

#### RAM

The `ram` column contains the memory capacity together with other information such as memory type and speed.

I will extract the RAM capacity in gigabytes.

In [51]:
df["ram"].head(10)

0      8 GB  LPDDR5 5500 Mhz RAM
1       16 GB  DDR5 4800 Mhz RAM
2       16 GB  DDR5 5600 Mhz RAM
3       16 GB  DDR5 5600 Mhz RAM
4       24 GB  DDR5 4800 Mhz RAM
5       16 GB  DDR5 5600 Mhz RAM
6         8 GB  zunifikowana RAM
7    24 GB  LPDDR5x 8533 Mhz RAM
8                32 GB  DDR5 RAM
9       16 GB  DDR4 3200 MHz RAM
Name: ram, dtype: object

In [52]:
df["ram"].str.extract(r"(\d+)")

,0
0,8
1,16
2,16
3,16
4,24
...,...
1070,16
1071,16
1072,32
1073,32


In [53]:
df["ram_gb"] = df["ram"].str.extract(r"(\d+)").astype(int)

In [54]:
df[["ram", "ram_gb"]].head(10)

,ram,ram_gb
0,8 GB LPDDR5 5500 Mhz RAM,8
1,16 GB DDR5 4800 Mhz RAM,16
2,16 GB DDR5 5600 Mhz RAM,16
3,16 GB DDR5 5600 Mhz RAM,16
4,24 GB DDR5 4800 Mhz RAM,24
5,16 GB DDR5 5600 Mhz RAM,16
6,8 GB zunifikowana RAM,8
7,24 GB LPDDR5x 8533 Mhz RAM,24
8,32 GB DDR5 RAM,32
9,16 GB DDR4 3200 MHz RAM,16


In [55]:
df["ram_gb"].value_counts().sort_index()

ram_gb
4        3
8       81
16     682
24      83
32     193
36       6
48       6
64      18
96       1
128      2
Name: count, dtype: int64

In [56]:
df["ram_gb"].isna().sum()

0

The RAM capacity was successfully extracted for all 1,075 laptops, with no missing values in the new `ram_gb` column.

#### Storage

The `storage` column contains the storage capacity and drive type.

I will extract the numeric storage capacity and inspect the units before converting the values into a standard format.

In [57]:
df["storage"].value_counts().head(20)

storage
512 GB SSD    618
1 TB SSD      357
256 GB SSD     42
2 TB SSD       38
512 GB          7
4 TB SSD        5
128 GB SSD      3
6 TB SSD        2
64 GB eMMC      2
512 TB SSD      1
Name: count, dtype: int64

In [58]:
df["storage"].unique()

array(['256 GB SSD', '512 GB SSD', '1 TB SSD', '512 GB', '128 GB SSD',
       '4 TB SSD', '2 TB SSD', '6 TB SSD', '512 TB SSD', '64 GB eMMC'],
      dtype=object)

The storage capacity is reported in both gigabytes (GB) and terabytes (TB), so the values need to be converted to a common unit.

I will standardize all storage capacities to gigabytes.

In [59]:
df[df["storage"] == "512 TB SSD"][["Title", "Price", "storage"]]

,Title,Price,storage
514,Laptop gamingowy ASUS TUF Gaming A14 FA401UM-R...,4399.0,512 TB SSD


The dataset contains one unusually large storage value: `512 TB SSD`.

This value is highly unusual for a laptop, so I will keep it for now and treat it as a potential data quality issue rather than removing it without further investigation.

In [60]:
df["storage"].str.extract(r"(\d+)")

,0
0,256
1,512
2,512
3,512
4,512
...,...
1070,512
1071,512
1072,1
1073,1


In [61]:
df["storage"].str.extract(r"(GB|TB)")

,0
0,GB
1,GB
2,GB
3,GB
4,GB
...,...
1070,GB
1071,GB
1072,TB
1073,TB


In [62]:
df["storage_value"] = (
    df["storage"]
    .str.extract(r"(\d+)")
    .astype(int)
)

df["storage_unit"] = (
    df["storage"]
    .str.extract(r"(GB|TB)")
)

In [63]:
df[["storage", "storage_value", "storage_unit"]].head(10)

,storage,storage_value,storage_unit
0,256 GB SSD,256,GB
1,512 GB SSD,512,GB
2,512 GB SSD,512,GB
3,512 GB SSD,512,GB
4,512 GB SSD,512,GB
5,512 GB SSD,512,GB
6,256 GB SSD,256,GB
7,512 GB SSD,512,GB
8,1 TB SSD,1,TB
9,512 GB,512,GB


In [64]:
df["storage_unit"].value_counts()

storage_unit
GB    672
TB    403
Name: count, dtype: int64

In [65]:
df["storage_value"].isna().sum()

0

In [66]:
df["storage_gb"] = df["storage_value"]

df.loc[df["storage_unit"] == "TB", "storage_gb"] *= 1024

In [67]:
df[["storage", "storage_value", "storage_unit", "storage_gb"]].head(10)

,storage,storage_value,storage_unit,storage_gb
0,256 GB SSD,256,GB,256
1,512 GB SSD,512,GB,512
2,512 GB SSD,512,GB,512
3,512 GB SSD,512,GB,512
4,512 GB SSD,512,GB,512
5,512 GB SSD,512,GB,512
6,256 GB SSD,256,GB,256
7,512 GB SSD,512,GB,512
8,1 TB SSD,1,TB,1024
9,512 GB,512,GB,512


In [68]:
df["storage_gb"].value_counts().sort_index()

storage_gb
64          2
128         3
256        42
512       625
1024      357
2048       38
4096        5
6144        2
524288      1
Name: count, dtype: int64

In [69]:
df[df["storage_gb"] > 10000][
    ["Title", "Price", "storage", "storage_gb"]
]

,Title,Price,storage,storage_gb
514,Laptop gamingowy ASUS TUF Gaming A14 FA401UM-R...,4399.0,512 TB SSD,524288


In [70]:
df = df.drop(columns=["storage_value", "storage_unit"])

In [71]:
df.columns

Index(['Title', 'Price', 'Image URL', 'screen', 'processor', 'ram', 'graphics',
       'storage', 'os', 'ram_gb', 'storage_gb'],
      dtype='object')

The storage values were successfully converted to gigabytes.

One potential data quality issue was identified: one laptop is listed with `512 TB SSD`, which corresponds to `524288 GB`. This value is highly unusual for a laptop, so I will keep the record for now and flag it as a potential data quality issue rather than removing it without further investigation.

#### Screen

The `screen` column contains several pieces of information, including screen size, resolution, and refresh rate.

I will extract these values into separate numeric columns.

In [72]:
df["screen"].value_counts().head(20)

screen
14 ",  1920 x 1200 pikseli 60 Hz                   107
16 ",  1920 x 1200 pikseli 60 Hz                   105
15,6 ",  1920 x 1080 pikseli 144 Hz                 73
15,6 ",  1920 x 1080 pikseli 60 Hz                  55
15,3",  2880 x 1864 pikseli                         38
13,6 ",  2560 x 1664 pikseli                        37
16 ",  2560 x 1600 pikseli 240 Hz                   35
15,6 ",  1920 x 1080 pikseli                        34
16 ",  1920 x 1200 pikseli                          34
16 ",  1920 x 1200 pikseli 165 Hz                   28
15,6 ",  1920 x 1080 pikseli 120 Hz                 27
16 ",  2560 x 1600 pikseli 180 Hz                   23
14 ", ekran dotykowy 2880 x 1800 pikseli 120 Hz     22
14 ", ekran dotykowy 1920 x 1200 pikseli 60 Hz      22
16 ",  1920 x 1200 pikseli 144 Hz                   21
14,2 ",  3024 x 1964 pikseli 120 Hz                 20
14 ",  1920 x 1200 pikseli                          19
17,3 ",  1920 x 1080 pikseli 144 Hz                 18
15,

In [73]:
df['screen'].str.extract(r'(\d+(?:\,\d+)?)')

,0
0,16
1,"15,6"
2,16
3,16
4,16
...,...
1070,"17,3"
1071,16
1072,14
1073,16


The screen size can use either a whole number or a decimal value with a comma, for example `16` or `15,6`.

The regular expression extracts the numeric part while supporting both formats.

In [74]:
df["screen_size"] = (
    df["screen"]
    .str.extract(r"(\d+(?:,\d+)?)")[0]
    .str.replace(",", ".", regex=False)
    .astype(float)
)

In [75]:
df[["screen", "screen_size"]].head(10)

,screen,screen_size
0,"16 "", 1920 x 1200 pikseli",16.0
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",15.6
2,"16 "", 2560 x 1600 pikseli 120 Hz",16.0
3,"16 "", 1920 x 1200 pikseli 144 Hz",16.0
4,"16 "", 1920 x 1200 pikseli 60 Hz",16.0
5,"15,3"", 2560 x 1600 pikseli 180 Hz",15.3
6,"13 "", 2408 x 1506 pikseli 60 Hz",13.0
7,"14 "", ekran dotykowy 1920 x 1200 pikseli 60 Hz",14.0
8,"16 "", 1920 x 1200 pikseli 60 Hz",16.0
9,"15,6 "", 1920 x 1080 pikseli",15.6


In [76]:
df["screen_size"].isna().sum()

0

In [77]:
df["screen_size"].min(), df["screen_size"].max()

(10.5, 18.0)

In [78]:
df["screen"].str.extract(r"(\d+)\s*x\s*(\d+)")

,0,1
0,1920,1200
1,1920,1080
2,2560,1600
3,1920,1200
4,1920,1200
...,...,...
1070,1920,1080
1071,2880,1800
1072,2880,1800
1073,2560,1600


In [79]:
df[["resolution_width", "resolution_height"]] = (
    df["screen"]
    .str.extract(r"(\d+)\s*x\s*(\d+)")
    .astype(int)
)

In [80]:
df[["screen", "resolution_width", "resolution_height"]].head(10)

,screen,resolution_width,resolution_height
0,"16 "", 1920 x 1200 pikseli",1920,1200
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",1920,1080
2,"16 "", 2560 x 1600 pikseli 120 Hz",2560,1600
3,"16 "", 1920 x 1200 pikseli 144 Hz",1920,1200
4,"16 "", 1920 x 1200 pikseli 60 Hz",1920,1200
5,"15,3"", 2560 x 1600 pikseli 180 Hz",2560,1600
6,"13 "", 2408 x 1506 pikseli 60 Hz",2408,1506
7,"14 "", ekran dotykowy 1920 x 1200 pikseli 60 Hz",1920,1200
8,"16 "", 1920 x 1200 pikseli 60 Hz",1920,1200
9,"15,6 "", 1920 x 1080 pikseli",1920,1080


The screen size and resolution were successfully extracted into numeric columns. No missing values were introduced during the extraction.

In [81]:
df["resolution_width"].min(), df["resolution_width"].max()

(1920, 3840)

In [82]:
df["resolution_height"].min(), df["resolution_height"].max()

(1080, 2400)

In [83]:
df["screen"].str.extract(r"(\d+)\s*Hz")

,0
0,NaN
1,144
2,120
3,144
4,60
...,...
1070,NaN
1071,120
1072,120
1073,240


In [84]:
df["screen"].str.extract(r"(\d+)\s*Hz").isna().sum()

0    229
dtype: int64

In [85]:
df["screen"].str.extract(r"(\d+)\s*Hz").value_counts()

60     361
120    154
144    130
165     99
240     54
180     32
90      11
360      3
300      2
Name: count, dtype: int64

The refresh rate is not specified for every laptop, so I will extract it separately and preserve missing values where the original specification does not provide this information.

In [86]:
df["refresh_rate"] = (
    df["screen"]
    .str.extract(r"(\d+)\s*Hz")
    .astype("Int64")
)

In [87]:
df[["screen", "refresh_rate"]].head(10)

,screen,refresh_rate
0,"16 "", 1920 x 1200 pikseli",<NA>
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",144
2,"16 "", 2560 x 1600 pikseli 120 Hz",120
3,"16 "", 1920 x 1200 pikseli 144 Hz",144
4,"16 "", 1920 x 1200 pikseli 60 Hz",60
5,"15,3"", 2560 x 1600 pikseli 180 Hz",180
6,"13 "", 2408 x 1506 pikseli 60 Hz",60
7,"14 "", ekran dotykowy 1920 x 1200 pikseli 60 Hz",60
8,"16 "", 1920 x 1200 pikseli 60 Hz",60
9,"15,6 "", 1920 x 1080 pikseli",<NA>


In [88]:
df[["screen_size", "resolution_width", "resolution_height", "refresh_rate"]].isna().sum()

screen_size            0
resolution_width       0
resolution_height      0
refresh_rate         229
dtype: int64

The refresh rate was extracted for 846 laptops. For 229 laptops, the refresh rate was not specified in the original data, so these values are kept as missing.

The resolution can be represented as the total number of pixels by multiplying the horizontal and vertical resolution.

This provides a single numeric feature that can be used in further analysis.

In [89]:
df["resolution_pixels"] = (
    df["resolution_width"] * df["resolution_height"]
)

In [90]:
df["resolution_pixels"].isna().sum()

0

In [91]:
df[["screen", "resolution_width", "resolution_height", "resolution_pixels"]].head(10)

,screen,resolution_width,resolution_height,resolution_pixels
0,"16 "", 1920 x 1200 pikseli",1920,1200,2304000
1,"15,6 "", 1920 x 1080 pikseli 144 Hz",1920,1080,2073600
2,"16 "", 2560 x 1600 pikseli 120 Hz",2560,1600,4096000
3,"16 "", 1920 x 1200 pikseli 144 Hz",1920,1200,2304000
4,"16 "", 1920 x 1200 pikseli 60 Hz",1920,1200,2304000
5,"15,3"", 2560 x 1600 pikseli 180 Hz",2560,1600,4096000
6,"13 "", 2408 x 1506 pikseli 60 Hz",2408,1506,3626448
7,"14 "", ekran dotykowy 1920 x 1200 pikseli 60 Hz",1920,1200,2304000
8,"16 "", 1920 x 1200 pikseli 60 Hz",1920,1200,2304000
9,"15,6 "", 1920 x 1080 pikseli",1920,1080,2073600


In [92]:
df["resolution_pixels"].value_counts().head(10)

resolution_pixels
2304000    390
2073600    284
4096000    148
5184000     72
5368320     38
4259840     37
5939136     20
9216000     17
3686400     14
7720704     14
Name: count, dtype: int64

#### Processor

The `processor` column contains processor brand, family, and model information.

There are 215 unique processor names in the dataset, and the naming conventions differ across manufacturers. For example, the dataset contains Intel Core, AMD Ryzen, Apple, Qualcomm, and MediaTek processors.

I will keep the original `processor` column unchanged and create separate analytical features from the cleaned values.

First, I will inspect the most common processor models.

In [93]:
df["processor"].value_counts().head(30)

processor
AMD Ryzen™ AI 7 350                 44
Apple M4                            42
AMD Ryzen™ 7 260                    40
Apple M5                            29
Intel® Core™ i5 13gen 13420H        27
Intel® Core™ Ultra 5 15gen 225U     27
Intel® Core™ i7 13gen 13620H        26
Intel® Core™ Ultra 9 15gen 275HX    21
Intel® Core™ Ultra 7 15gen 258V     20
Intel® Core™ Ultra 7 15gen 255U     20
Intel® Core™ i5 13gen 1334U         19
Intel® Core™ 5 14gen 120U           19
AMD Ryzen™ AI 9 365                 18
Intel® Core™ i5 12gen 1235U         18
AMD Ryzen™ AI 7 445                 18
Intel® Core™ i7 14gen 14650HX       16
Intel® Core™ i5 11gen 1135G7        14
AMD Ryzen™ AI 5 340                 14
Intel® Core™ Ultra 7 15gen 256V     12
Intel® Core™ Ultra 7 15gen 255H     12
AMD Ryzen™ 5 5600H                  12
Intel® Core™ i5 12gen 12450H        11
Intel® Core™ 7 15gen 240H           10
Intel® Core™ i5 13gen 1335U         10
Intel® Core™ Ultra 9 15gen 288V     10
Intel® Core™ Ul

The dataset contains many different processor models, including several models that appear only once. Therefore, using the full processor name as a categorical feature would result in a high-cardinality column.

I will first standardize the processor names and extract the manufacturer as a separate feature.

In [94]:
df["cpu_clean"] = (
    df["processor"]
    .str.replace(r"[®™]", "", regex=True)
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

In [95]:
df[["processor", "cpu_clean"]].head(10)

,processor,cpu_clean
0,AMD Ryzen™ 5 40,AMD Ryzen 5 40
1,Intel® Core™ i5 12gen 12600HX,Intel Core i5 12gen 12600HX
2,Intel® Core™ 5 15gen 210H,Intel Core 5 15gen 210H
3,Intel® Core™ 5 210H,Intel Core 5 210H
4,AMD Ryzen™ 5 7535HS,AMD Ryzen 5 7535HS
5,AMD Ryzen™ 5 240,AMD Ryzen 5 240
6,Apple A18 Pro,Apple A18 Pro
7,AMD Ryzen™ AI 7 445,AMD Ryzen AI 7 445
8,Intel® Core™ 5 14gen 120U,Intel Core 5 14gen 120U
9,AMD Ryzen™ 7 5825U,AMD Ryzen 7 5825U


Next, I will extract the processor manufacturer from the cleaned processor name.

In [96]:
df["cpu_brand"] = df["cpu_clean"].str.split().str[0]

In [97]:
df["cpu_brand"].value_counts()

cpu_brand
Intel       604
AMD         333
Apple       121
Qualcomm     16
MediaTek      1
Name: count, dtype: int64

In [98]:
df["cpu_brand"].isna().sum()

0

In [99]:
df.groupby("cpu_brand")["processor"].nunique().sort_values(ascending=False)

cpu_brand
Intel       127
AMD          67
Apple        11
Qualcomm      9
MediaTek      1
Name: processor, dtype: int64

In [100]:
df.loc[df["cpu_brand"] == "Intel", "cpu_clean"].head(30)

1           Intel Core i5 12gen 12600HX
2               Intel Core 5 15gen 210H
3                     Intel Core 5 210H
8               Intel Core 5 14gen 120U
10          Intel Core i5 14gen 14450HX
12           Intel Core i5 13gen 13420H
14          Intel Core i7 13gen 13650HX
16       Intel Core Ultra 7 15gen 255HX
18           Intel Core i5 13gen 13420H
21        Intel Core Ultra 5 14gen 135H
23        Intel Core Ultra 5 14gen 226V
26          Intel Core i5 12gen 12450HX
28          Intel Core i7 14gen 14650HX
31        Intel Core Ultra 5 15gen 225U
33              Intel Core 5 14gen 120U
35          Intel Core i5 13gen 13450HX
45              Intel Core 5 14gen 120U
47       Intel Core Ultra 7 15gen 255HX
54          Intel Core i7 14gen 14650HX
56          Intel Core i7 14gen 14650HX
58           Intel Core i7 13gen 13620H
59       Intel Core Ultra 9 15gen 275HX
60        Intel Core Ultra 5 15gen 225U
62                           Intel N150
63              Intel Core 3 14gen 100U


In [101]:
df.loc[df["cpu_brand"] == "AMD", "cpu_clean"].head(30)

0         AMD Ryzen 5 40
4     AMD Ryzen 5 7535HS
5        AMD Ryzen 5 240
7     AMD Ryzen AI 7 445
9      AMD Ryzen 7 5825U
13       AMD Ryzen 7 250
19       AMD Ryzen 7 250
20     AMD Ryzen 5 7530U
22     AMD Ryzen 7 7730U
24    AMD Ryzen 9 8940HX
25       AMD Ryzen 5 150
27    AMD Ryzen 5 7235HS
29    AMD Ryzen AI 5 340
30    AMD Ryzen 5 8645HS
34    AMD Ryzen AI 5 340
36       AMD Ryzen 7 260
37       AMD Ryzen 7 260
38    AMD Ryzen AI 7 350
39       AMD Ryzen 5 240
43    AMD Ryzen AI 5 340
44    AMD Ryzen AI 5 430
46       AMD Ryzen 7 260
48       AMD Ryzen 5 150
49       AMD Ryzen 7 260
50       AMD Ryzen 5 150
51    AMD Ryzen 9 9955HX
53    AMD Ryzen AI 7 350
55    AMD Ryzen AI 7 350
66    AMD Ryzen AI 7 445
67    AMD Ryzen 9 9955HX
Name: cpu_clean, dtype: object

In [102]:
df.loc[df["cpu_brand"] == "Apple", "cpu_clean"].head(20)

6      Apple A18 Pro
15          Apple M4
17     Apple A18 Pro
32          Apple M5
40          Apple M3
41     Apple A18 Pro
52     Apple A18 Pro
57          Apple M5
61          Apple M4
65      Apple M5 Pro
68     Apple A18 Pro
82          Apple M5
99     Apple A18 Pro
114    Apple A18 Pro
116     Apple M5 Pro
121         Apple M5
128         Apple M4
130         Apple M5
133     Apple M4 Pro
135     Apple M4 Max
Name: cpu_clean, dtype: object

In [103]:
df.loc[df["cpu_brand"] == "Qualcomm", "cpu_clean"].head(20)

42      Qualcomm Snapdragon X2 Elite Extreme X2E94100
100           Qualcomm Snapdragon X2 Elite X2E-88-100
117             Qualcomm Snapdragon X Plus X1P-42-100
145           Qualcomm Snapdragon X2 Elite X2E-88-100
172                   Qualcomm Snapdragon X X1-26-100
213             Qualcomm Snapdragon X2 Elite X2E88100
585            Qualcomm Snapdragon X Elite X1E-78-100
688                        Qualcomm Snapdragon X Plus
702            Qualcomm Snapdragon X Elite X1E-78-100
703                  Qualcomm Snapdragon X X1-26 -100
731            Qualcomm Snapdragon X Elite X1E-78-100
749                        Qualcomm Snapdragon X Plus
800             Qualcomm Snapdragon X Plus X1P-42-100
831                   Qualcomm Snapdragon X X1-26-100
960                   Qualcomm Snapdragon X X1-26-100
1020                  Qualcomm Snapdragon X X1-26-100
Name: cpu_clean, dtype: object

In [104]:
df["cpu_clean"].value_counts().tail(50)

cpu_clean
Intel Core i7 14gen 14700HX              1
Intel Core i3 N305                       1
Intel Core Ultra 5 14gen 135H            1
Intel Core Ultra 7 14gen 258V            1
Intel Core i7 1185G7                     1
Intel Core i7 11gen 1195G7               1
MediaTek Kompanio 520                    1
Apple M1 Pro                             1
AMD Ryzen AI 7 360                       1
AMD Ryzen 9 270                          1
Intel N150                               1
AMD Ryzen 7 8840U                        1
Intel Core i7 10gen 10510U               1
Intel Core 7 Series 3 350                1
Intel Core Ultra 7 14gen 255H            1
Intel Core Ultra 7 Series 2 251HX        1
Intel Core i7 13gen 13700H               1
AMD Ryzen 7 8745HX                       1
Intel Core Ultra 9 14gen 275HX           1
Intel Core i3 12gen 1215U                1
Intel Core Ultra 7 15gen 265U            1
Intel Core Ultra 7 14gen 256V            1
Intel Core 5 Series 3 320                1
I

In [105]:
df.loc[
    ~df["cpu_brand"].isin(["Intel", "AMD", "Apple", "Qualcomm", "MediaTek"]),
    ["processor", "cpu_clean", "cpu_brand"]
]

,processor,cpu_clean,cpu_brand


##### Intel

In [106]:
df.loc[df["cpu_brand"] == "Intel", "cpu_clean"].str.extract(
    r"(Core Ultra [3579]|Core i[3579]|Core [3579]|N\d+)"
)[0].value_counts(dropna=False)

0
Core i5         148
Core Ultra 7    111
Core i7         104
Core Ultra 5     84
Core Ultra 9     54
Core 5           43
Core 7           24
Core i3           9
NaN               8
Core 9            7
Core 3            5
Core i9           5
N150              1
N4500             1
Name: count, dtype: int64

In [107]:
df.loc[
    (df["cpu_brand"] == "Intel") &
    ~df["cpu_clean"].str.contains(
        r"Core Ultra [3579]|Core i[3579]|Core [3579]|N\d+",
        regex=True
    ),
    ["processor", "cpu_clean"]
]

,processor,cpu_clean
76,Intel® Core™ Ultra X7 Series 3 358H,Intel Core Ultra X7 Series 3 358H
91,Intel® Core™ Ultra X9 Series 3 388H,Intel Core Ultra X9 Series 3 388H
220,Intel® Core™ Ultra X9 Series 3 378H,Intel Core Ultra X9 Series 3 378H
310,Intel® Core™ Ultra X7 Series 3 358H,Intel Core Ultra X7 Series 3 358H
509,Intel® Core™ Ultra X9 Series 3 388H,Intel Core Ultra X9 Series 3 388H
510,Intel® Core™ Ultra X7 Series 3 358H,Intel Core Ultra X7 Series 3 358H
945,Intel® Pentium® Gold 6500Y,Intel Pentium Gold 6500Y
959,Intel® Pentium® Gold 6500Y,Intel Pentium Gold 6500Y


In [108]:
intel_mask = df["cpu_brand"] == "Intel"

df.loc[intel_mask, "cpu_family"] = (
    df.loc[intel_mask, "cpu_clean"]
    .str.extract(
        r"(Core Ultra [3579]|Core Ultra X[79]|Core i[3579]|Core [3579]|N\d+|Pentium Gold)"
    )[0]
)

In [109]:
df.loc[intel_mask, "cpu_family"].value_counts(dropna=False)

cpu_family
Core i5          148
Core Ultra 7     111
Core i7          104
Core Ultra 5      84
Core Ultra 9      54
Core 5            43
Core 7            24
Core i3            9
Core 9             7
Core 3             5
Core i9            5
Core Ultra X7      3
Core Ultra X9      3
Pentium Gold       2
N150               1
N4500              1
Name: count, dtype: int64

In [110]:
df.loc[
    intel_mask & df["cpu_family"].isna(),
    ["processor", "cpu_clean"]
]

,processor,cpu_clean


##### AMD

In [111]:
df.loc[df["cpu_brand"] == "AMD", "cpu_clean"].str.extract(
    r"(Ryzen AI [3579]|Ryzen [3579])"
)[0].value_counts(dropna=False)

0
Ryzen 7       103
Ryzen 5        87
Ryzen AI 7     70
Ryzen AI 5     30
Ryzen AI 9     26
Ryzen 9        11
NaN             4
Ryzen 3         2
Name: count, dtype: int64

In [112]:
df.loc[
    (df["cpu_brand"] == "AMD") &
    ~df["cpu_clean"].str.contains(
        r"Ryzen AI [3579]|Ryzen [3579]",
        regex=True
    ),
    ["processor", "cpu_clean"]
]

,processor,cpu_clean
96,AMD Ryzen™ AI Max+ 395,AMD Ryzen AI Max+ 395
140,AMD Ryzen™ AI Max+ 388,AMD Ryzen AI Max+ 388
455,AMD Ryzen™ AI Max+ 395,AMD Ryzen AI Max+ 395
854,AMD Ryzen™ Al Max+ 395,AMD Ryzen Al Max+ 395


In [113]:
amd_mask = df["cpu_brand"] == "AMD"

df.loc[amd_mask, "cpu_family"] = (
    df.loc[amd_mask, "cpu_clean"]
    .str.extract(
        r"(Ryzen AI Max\+|Ryzen AI [3579]|Ryzen [3579])"
    )[0]
)

In [114]:
df.loc[amd_mask, "cpu_family"].value_counts(dropna=False)

cpu_family
Ryzen 7          103
Ryzen 5           87
Ryzen AI 7        70
Ryzen AI 5        30
Ryzen AI 9        26
Ryzen 9           11
Ryzen AI Max+      3
Ryzen 3            2
NaN                1
Name: count, dtype: int64

In [115]:
df.loc[
    amd_mask & df["cpu_family"].isna(),
    ["processor", "cpu_clean"]
]

,processor,cpu_clean
854,AMD Ryzen™ Al Max+ 395,AMD Ryzen Al Max+ 395


In [116]:
df["cpu_clean"] = df["cpu_clean"].str.replace(
    "Ryzen Al Max+",
    "Ryzen AI Max+",
    regex=False
)

In [117]:
df.loc[amd_mask, "cpu_family"] = (
    df.loc[amd_mask, "cpu_clean"]
    .str.extract(
        r"(Ryzen AI Max\+|Ryzen AI [3579]|Ryzen [3579])"
    )[0]
)

In [118]:
df.loc[
    amd_mask & df["cpu_family"].isna(),
    ["processor", "cpu_clean"]
]

,processor,cpu_clean


##### Apple

In [119]:
apple_mask = df["cpu_brand"] == "Apple"

df.loc[apple_mask, "cpu_family"] = (
    df.loc[apple_mask, "cpu_clean"]
    .str.replace("Apple ", "", regex=False)
)

In [120]:
df.loc[apple_mask, "cpu_family"].value_counts(dropna=False)

cpu_family
M4         42
M5         29
M5 Pro      9
A18 Pro     8
M3          7
M2          7
M4 Pro      6
M5 Max      6
M4 Max      3
M1          3
M1 Pro      1
Name: count, dtype: int64

##### Qualcomm

In [121]:
qualcomm_mask = df["cpu_brand"] == "Qualcomm"

df.loc[qualcomm_mask, "cpu_family"] = (
    df.loc[qualcomm_mask, "cpu_clean"]
    .str.extract(
        r"(Snapdragon X2 Elite Extreme|Snapdragon X2 Elite|Snapdragon X Elite|Snapdragon X Plus|Snapdragon X)"
    )[0]
)

In [122]:
df.loc[qualcomm_mask, "cpu_family"].value_counts(dropna=False)

cpu_family
Snapdragon X                   5
Snapdragon X Plus              4
Snapdragon X2 Elite            3
Snapdragon X Elite             3
Snapdragon X2 Elite Extreme    1
Name: count, dtype: int64

##### MediaTek

In [123]:
mediatek_mask = df["cpu_brand"] == "MediaTek"

df.loc[mediatek_mask, "cpu_family"] = (
    df.loc[mediatek_mask, "cpu_clean"]
    .str.replace("MediaTek ", "", regex=False)
)

In [124]:
df.loc[mediatek_mask, "cpu_family"].value_counts(dropna=False)

cpu_family
Kompanio 520    1
Name: count, dtype: int64

In [125]:
df["cpu_family"].isna().sum()

0

In [126]:
df["cpu_family"].value_counts(dropna=False)

cpu_family
Core i5                        148
Core Ultra 7                   111
Core i7                        104
Ryzen 7                        103
Ryzen 5                         87
Core Ultra 5                    84
Ryzen AI 7                      70
Core Ultra 9                    54
Core 5                          43
M4                              42
Ryzen AI 5                      30
M5                              29
Ryzen AI 9                      26
Core 7                          24
Ryzen 9                         11
Core i3                          9
M5 Pro                           9
A18 Pro                          8
M3                               7
M2                               7
Core 9                           7
M4 Pro                           6
M5 Max                           6
Snapdragon X                     5
Core i9                          5
Core 3                           5
Ryzen AI Max+                    4
Snapdragon X Plus                4
Snapdrago

#### Generation

In [127]:
df["cpu_generation"] = pd.Series(pd.NA, index=df.index, dtype="Int64")

df.loc[intel_mask, "cpu_generation"] = (
    df.loc[intel_mask, "cpu_clean"]
    .str.extract(r"(\d+)(?=gen)")[0]
    .astype("Int64")
)

In [128]:
df["cpu_generation"].value_counts(dropna=False)

cpu_generation
<NA>    542
15      200
13      109
14       94
12       66
11       45
10       14
9         3
8         2
Name: count, dtype: Int64

In [129]:
df.loc[
    intel_mask & df["cpu_generation"].isna(),
    ["processor", "cpu_clean", "cpu_family"]
]

,processor,cpu_clean,cpu_family
3,Intel® Core™ 5 210H,Intel Core 5 210H,Core 5
62,Intel® N150,Intel N150,N150
76,Intel® Core™ Ultra X7 Series 3 358H,Intel Core Ultra X7 Series 3 358H,Core Ultra X7
81,Intel® Core™ 5 210H,Intel Core 5 210H,Core 5
91,Intel® Core™ Ultra X9 Series 3 388H,Intel Core Ultra X9 Series 3 388H,Core Ultra X9
...,...,...,...
971,Intel® Core™ Ultra 7 Series 3 355,Intel Core Ultra 7 Series 3 355,Core Ultra 7
1007,Intel® Core™ i3 N305,Intel Core i3 N305,Core i3
1017,Intel® Core™ 5 120U,Intel Core 5 120U,Core 5
1028,Intel® Core™ Ultra 9 Series 3 386H,Intel Core Ultra 9 Series 3 386H,Core Ultra 9


The processor generation was extracted for Intel processors where the original name explicitly contains a generation marker such as `12gen`, `13gen`, `14gen`, or `15gen`.

Some Intel processor names do not contain an explicit `gen` marker, and processors from AMD, Apple, Qualcomm, and MediaTek use different naming conventions. These cases are therefore kept as missing values rather than assigning potentially incorrect generations.

In [130]:
df[["processor", "cpu_clean", "cpu_brand", "cpu_family", "cpu_generation"]].head(20)

,processor,cpu_clean,cpu_brand,cpu_family,cpu_generation
0,AMD Ryzen™ 5 40,AMD Ryzen 5 40,AMD,Ryzen 5,<NA>
1,Intel® Core™ i5 12gen 12600HX,Intel Core i5 12gen 12600HX,Intel,Core i5,12
2,Intel® Core™ 5 15gen 210H,Intel Core 5 15gen 210H,Intel,Core 5,15
3,Intel® Core™ 5 210H,Intel Core 5 210H,Intel,Core 5,<NA>
4,AMD Ryzen™ 5 7535HS,AMD Ryzen 5 7535HS,AMD,Ryzen 5,<NA>
5,AMD Ryzen™ 5 240,AMD Ryzen 5 240,AMD,Ryzen 5,<NA>
6,Apple A18 Pro,Apple A18 Pro,Apple,A18 Pro,<NA>
7,AMD Ryzen™ AI 7 445,AMD Ryzen AI 7 445,AMD,Ryzen AI 7,<NA>
8,Intel® Core™ 5 14gen 120U,Intel Core 5 14gen 120U,Intel,Core 5,14
9,AMD Ryzen™ 7 5825U,AMD Ryzen 7 5825U,AMD,Ryzen 7,<NA>


In [131]:
df[["cpu_brand", "cpu_family", "cpu_generation"]].isna().sum()

cpu_brand           0
cpu_family          0
cpu_generation    542
dtype: int64

#### Result

The original `processor` column was preserved, while a cleaned `cpu_clean` column was created by removing trademark symbols and normalizing whitespace.

The processor manufacturer was extracted into `cpu_brand`. The dataset contains five manufacturers: Intel, AMD, Apple, Qualcomm, and MediaTek.

The main processor family was extracted into `cpu_family`, with manufacturer-specific naming conventions taken into account.

For Intel processors, the generation was extracted into `cpu_generation` when an explicit `gen` marker was present in the original processor name. Missing generation values were preserved for processors where the generation could not be reliably determined from the available text.

No missing values were introduced in `cpu_brand` or `cpu_family`.

#### Graphics

The `graphics` column contains information about the laptop's graphics processor.

The values use different formats and may contain one or multiple graphics processors. For example, some laptops have only integrated graphics, while others contain a dedicated NVIDIA GPU together with an integrated Intel or AMD GPU.

I will clean the text and create analytical features while preserving the original `graphics` column unchanged.

In [132]:
df["graphics"].head(20)

0                         AMD  Radeon™ 610M Graphics
1     NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics
2         NVIDIA® GeForce RTX™ 3050 + Intel Graphics
3         NVIDIA® GeForce RTX™ 3050 + Intel Graphics
4                         AMD  Radeon™ 660M Graphics
5       NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 760M
6                          Apple  A18 Pro (5-rdzeni)
7                                  AMD  Radeon™ 840M
8                                   Intel®  Graphics
9                              AMD  Radeon™ Graphics
10    NVIDIA® GeForce RTX™ 5050 + Intel UHD Graphics
11                                    ARM® Mali G-52
12    NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics
13      NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 780M
14    NVIDIA® GeForce RTX™ 5060 + Intel UHD Graphics
15                             Apple  M4 (10-rdzeni)
16        NVIDIA® GeForce RTX™ 5060 + Intel Graphics
17                         Apple  A18 Pro (5-rdzeni)
18    NVIDIA® GeForce RTX™ 5050 + Intel UHD Gr

In [133]:
df["graphics"].value_counts().head(30)

graphics
Intel®  Graphics                                                  146
Intel®  Iris Xe Graphics                                           64
Intel®  UHD Graphics                                               64
AMD  Radeon™ Graphics                                              41
Apple  M4 (10-rdzeni)                                              38
AMD  Radeon™ 840M                                                  30
Apple  M5 (10-rdzeni)                                              25
Intel®  Arc Graphics                                               23
NVIDIA® GeForce RTX™ 5060 + Intel UHD Graphics                     16
AMD  Radeon™ 860M                                                  15
Intel®  Iris Xe Graphics (działający jako Intel® UHD Graphics)     13
NVIDIA® GeForce RTX™ 5060 + Intel Graphics                         13
AMD  Radeon™ 860M Graphics                                         12
NVIDIA® GeForce RTX 5060 + AMD Radeon™ 780M                        11
NVIDIA® GeF

In [134]:
df["graphics"].nunique()

205

In [135]:
df["graphics"].isna().sum()

0

The column contains 205 unique graphics configurations and has no missing values.

Because several rows contain multiple GPUs separated by `+`, I will preserve the complete configuration and create additional standardized features for analysis.

In [136]:
df["gpu_clean"] = (
    df["graphics"]
    .str.replace(r"[®™]", "", regex=True)
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

In [137]:
df[["graphics", "gpu_clean"]].head(20)

,graphics,gpu_clean
0,AMD Radeon™ 610M Graphics,AMD Radeon 610M Graphics
1,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics
2,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics
3,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics
4,AMD Radeon™ 660M Graphics,AMD Radeon 660M Graphics
5,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 760M,NVIDIA GeForce RTX 5050 + AMD Radeon 760M
6,Apple A18 Pro (5-rdzeni),Apple A18 Pro (5-rdzeni)
7,AMD Radeon™ 840M,AMD Radeon 840M
8,Intel® Graphics,Intel Graphics
9,AMD Radeon™ Graphics,AMD Radeon Graphics


In [138]:
df["gpu_brand"] = df["gpu_clean"].str.split().str[0]

In [139]:
df["gpu_brand"] = pd.NA

df.loc[
    df["gpu_clean"].str.contains("NVIDIA", case=False, na=False),
    "gpu_brand"
] = "NVIDIA"

df.loc[
    df["gpu_clean"].str.contains("AMD", case=False, na=False) &
    df["gpu_brand"].isna(),
    "gpu_brand"
] = "AMD"

df.loc[
    df["gpu_clean"].str.contains("Intel", case=False, na=False) &
    df["gpu_brand"].isna(),
    "gpu_brand"
] = "Intel"

df.loc[
    df["gpu_clean"].str.contains("Apple", case=False, na=False) &
    df["gpu_brand"].isna(),
    "gpu_brand"
] = "Apple"

df.loc[
    df["gpu_clean"].str.contains("Qualcomm", case=False, na=False) &
    df["gpu_brand"].isna(),
    "gpu_brand"
] = "Qualcomm"

df.loc[
    df["gpu_clean"].str.contains("ARM", case=False, na=False) &
    df["gpu_brand"].isna(),
    "gpu_brand"
] = "ARM"

In [140]:
df["gpu_brand"].value_counts(dropna=False)

gpu_brand
Intel       402
NVIDIA      351
AMD         184
Apple       121
Qualcomm     16
ARM           1
Name: count, dtype: int64

In [141]:
df.loc[
    df["gpu_brand"].isna(),
    ["graphics", "gpu_clean"]
]

,graphics,gpu_clean


No rows are returned, so no missing values were introduced in `gpu_brand`.

In [142]:
df["has_dedicated_gpu"] = (
    df["gpu_clean"].str.contains(
        r"NVIDIA\s+(?:GeForce\s+)?(?:RTX|GTX|MX)|"
        r"AMD\s+Radeon\s+RX",
        case=False,
        na=False
    )
)

In [143]:
df["has_dedicated_gpu"].value_counts()

has_dedicated_gpu
False    722
True     353
Name: count, dtype: int64

Therefore, 353 laptops contain a dedicated GPU, while 722 do not.

In [144]:
df.loc[
    df["has_dedicated_gpu"],
    ["graphics", "gpu_clean", "has_dedicated_gpu"]
].head(20)

,graphics,gpu_clean,has_dedicated_gpu
1,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics,True
2,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,True
3,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,True
5,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 760M,NVIDIA GeForce RTX 5050 + AMD Radeon 760M,True
10,NVIDIA® GeForce RTX™ 5050 + Intel UHD Graphics,NVIDIA GeForce RTX 5050 + Intel UHD Graphics,True
12,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics,True
13,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 780M,NVIDIA GeForce RTX 5050 + AMD Radeon 780M,True
14,NVIDIA® GeForce RTX™ 5060 + Intel UHD Graphics,NVIDIA GeForce RTX 5060 + Intel UHD Graphics,True
16,NVIDIA® GeForce RTX™ 5060 + Intel Graphics,NVIDIA GeForce RTX 5060 + Intel Graphics,True
18,NVIDIA® GeForce RTX™ 5050 + Intel UHD Graphics,NVIDIA GeForce RTX 5050 + Intel UHD Graphics,True


##### NVIDIA GPU model

In [145]:
df["nvidia_gpu"] = (
    df["gpu_clean"]
    .str.replace(r"Geforce", "GeForce", case=False, regex=True)
    .str.extract(
        r"((?:GeForce\s+)?(?:RTX|GTX|MX)\s*\d+(?:\s*Ti)?(?:\s*Max-Q)?)",
        expand=False
    )
)

In [146]:
df["nvidia_gpu"] = (
    df["nvidia_gpu"]
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

In [147]:
df.loc[
    df["gpu_clean"].str.contains("NVIDIA", case=False, na=False),
    ["graphics", "gpu_clean", "nvidia_gpu"]
].head(30)

,graphics,gpu_clean,nvidia_gpu
1,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics,GeForce RTX 4050
2,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,GeForce RTX 3050
3,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,GeForce RTX 3050
5,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 760M,NVIDIA GeForce RTX 5050 + AMD Radeon 760M,GeForce RTX 5050
10,NVIDIA® GeForce RTX™ 5050 + Intel UHD Graphics,NVIDIA GeForce RTX 5050 + Intel UHD Graphics,GeForce RTX 5050
12,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics,GeForce RTX 4050
13,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 780M,NVIDIA GeForce RTX 5050 + AMD Radeon 780M,GeForce RTX 5050
14,NVIDIA® GeForce RTX™ 5060 + Intel UHD Graphics,NVIDIA GeForce RTX 5060 + Intel UHD Graphics,GeForce RTX 5060
16,NVIDIA® GeForce RTX™ 5060 + Intel Graphics,NVIDIA GeForce RTX 5060 + Intel Graphics,GeForce RTX 5060
18,NVIDIA® GeForce RTX™ 5050 + Intel UHD Graphics,NVIDIA GeForce RTX 5050 + Intel UHD Graphics,GeForce RTX 5050


In [148]:
df.loc[
    df["gpu_clean"].str.contains("NVIDIA", case=False, na=False)
    & df["nvidia_gpu"].isna(),
    ["graphics", "gpu_clean"]
]

,graphics,gpu_clean


No rows are returned. Therefore, all NVIDIA configurations in the dataset have a successfully extracted NVIDIA model.

In [149]:
df["gpu_model"] = df["nvidia_gpu"]

##### AMD

In [150]:
amd_mask = (
    df["gpu_brand"].eq("AMD")
    & df["gpu_model"].isna()
)

df.loc[amd_mask, "gpu_model"] = (
    df.loc[amd_mask, "gpu_clean"]
    .str.extract(
        r"(Radeon\s+(?:RX\s+)?[A-Za-z0-9]+)",
        expand=False
    )
)

##### Intel

In [151]:
intel_mask = (
    df["gpu_brand"].eq("Intel")
    & df["gpu_model"].isna()
)

df.loc[intel_mask, "gpu_model"] = (
    df.loc[intel_mask, "gpu_clean"]
    .str.extract(
        r"(Arc(?:\s+Graphics)?(?:\s+\d+[A-Z])?|"
        r"Iris Xe Graphics|"
        r"Iris Plus Graphics|"
        r"UHD Graphics(?:\s+\d+)?|"
        r"Graphics)",
        expand=False
    )
)

##### Apple

In [152]:
apple_mask = (
    df["gpu_brand"].eq("Apple")
    & df["gpu_model"].isna()
)

df.loc[apple_mask, "gpu_model"] = (
    df.loc[apple_mask, "gpu_clean"]
    .str.extract(
        r"(M[1-5](?:\s+(?:Pro|Max))?|A18 Pro)",
        expand=False
    )
)

##### Qualcomm

In [153]:
qualcomm_mask = (
    df["gpu_brand"].eq("Qualcomm")
    & df["gpu_model"].isna()
)

df.loc[qualcomm_mask, "gpu_model"] = "Adreno"

##### ARM

In [154]:
arm_mask = (
    df["gpu_brand"].eq("ARM")
    & df["gpu_model"].isna()
)

df.loc[arm_mask, "gpu_model"] = (
    df.loc[arm_mask, "gpu_clean"]
    .str.replace("ARM ", "", regex=False)
)

Finally, I will remove the redundant Graphics suffix from the extracted model names.

In [155]:
df["gpu_model"] = (
    df["gpu_model"]
    .str.replace(r"\s*Graphics$", "", regex=True)
    .str.strip()
    .replace("", pd.NA)
)

In [156]:
df["gpu_model"].value_counts().head(30)

gpu_model
Iris Xe                79
GeForce RTX 5060       75
UHD                    65
Arc                    45
GeForce RTX 5070       43
GeForce RTX 5050       42
M4                     42
Radeon 840M            41
Radeon                 41
GeForce RTX 4050       30
GeForce RTX 3050       30
M5                     29
Radeon 860M            27
GeForce RTX 4060       22
GeForce RTX 5070 Ti    21
Arc Graphics 140V      18
GeForce RTX 5080       18
Radeon 660M            16
Adreno                 16
Arc 140V               15
GeForce RTX 3060       14
Radeon 780M            13
Radeon 740M             9
Arc Graphics 130V       9
GeForce RTX 3050 Ti     9
M5 Pro                  9
GeForce RTX 5090        9
A18 Pro                 8
Arc 140T                8
M2                      7
Name: count, dtype: int64

##### GPU manufacturer indicators

In addition to gpu_brand, I will create binary indicators showing whether the graphics configuration contains NVIDIA, AMD, or Intel graphics.

These features are useful because a laptop may contain more than one GPU. For example, a configuration such as NVIDIA GeForce RTX 5060 + Intel UHD Graphics belongs to NVIDIA and Intel simultaneously.

In [157]:
df["has_nvidia_gpu"] = df["gpu_clean"].str.contains(
    "NVIDIA",
    case=False,
    na=False
)

df["has_amd_gpu"] = df["gpu_clean"].str.contains(
    "AMD",
    case=False,
    na=False
)

df["has_intel_gpu"] = df["gpu_clean"].str.contains(
    "Intel",
    case=False,
    na=False
)

In [158]:
df[
    [
        "graphics",
        "gpu_clean",
        "gpu_brand",
        "gpu_model",
        "nvidia_gpu",
        "has_dedicated_gpu",
        "has_nvidia_gpu",
        "has_amd_gpu",
        "has_intel_gpu"
    ]
].head(50)

,graphics,gpu_clean,gpu_brand,gpu_model,nvidia_gpu,has_dedicated_gpu,has_nvidia_gpu,has_amd_gpu,has_intel_gpu
0,AMD Radeon™ 610M Graphics,AMD Radeon 610M Graphics,AMD,Radeon 610M,NaN,False,False,True,False
1,NVIDIA® GeForce RTX™ 4050 + Intel UHD Graphics,NVIDIA GeForce RTX 4050 + Intel UHD Graphics,NVIDIA,GeForce RTX 4050,GeForce RTX 4050,True,True,False,True
2,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,NVIDIA,GeForce RTX 3050,GeForce RTX 3050,True,True,False,True
3,NVIDIA® GeForce RTX™ 3050 + Intel Graphics,NVIDIA GeForce RTX 3050 + Intel Graphics,NVIDIA,GeForce RTX 3050,GeForce RTX 3050,True,True,False,True
4,AMD Radeon™ 660M Graphics,AMD Radeon 660M Graphics,AMD,Radeon 660M,NaN,False,False,True,False
5,NVIDIA® GeForce RTX™ 5050 + AMD Radeon™ 760M,NVIDIA GeForce RTX 5050 + AMD Radeon 760M,NVIDIA,GeForce RTX 5050,GeForce RTX 5050,True,True,True,False
6,Apple A18 Pro (5-rdzeni),Apple A18 Pro (5-rdzeni),Apple,A18 Pro,NaN,False,False,False,False
7,AMD Radeon™ 840M,AMD Radeon 840M,AMD,Radeon 840M,NaN,False,False,True,False
8,Intel® Graphics,Intel Graphics,Intel,<NA>,NaN,False,False,False,True
9,AMD Radeon™ Graphics,AMD Radeon Graphics,AMD,Radeon,NaN,False,False,True,False


#### Result

The original `graphics` column was preserved unchanged, while a cleaned `gpu_clean` column was created by removing trademark symbols and normalizing whitespace.

The `graphics` column contains 205 unique GPU configurations and has no missing values. Some configurations contain multiple GPUs separated by `+`, so the original configuration was preserved while additional features were created for analysis.

The GPU manufacturer was extracted into `gpu_brand`. Six manufacturers are represented in the dataset: Intel, NVIDIA, AMD, Apple, Qualcomm, and ARM. No missing values were introduced in this feature.

A `has_dedicated_gpu` feature was created to identify laptops containing a dedicated GPU. According to the applied classification, 353 laptops contain a dedicated GPU, while 722 do not.

For NVIDIA configurations, the specific GPU model was extracted into `nvidia_gpu`. A unified `gpu_model` feature was then created for the different GPU manufacturers. When the source text did not contain a specific model, the value was kept missing rather than assigning an unsupported model. For example, Intel Graphics does not provide enough information to determine a specific GPU model.

Finally, binary indicators `has_nvidia_gpu`, `has_amd_gpu`, and `has_intel_gpu` were created to identify the presence of each manufacturer independently. This is particularly useful for configurations containing multiple GPUs.